wgpu-core/src/command/bundle.rs

```diff
@@ -0,0 +1,936 @@
+                    None
+                }
+            })
+    }
+}
+
+impl<G: GlobalIdentityHandlerFactory> Global<G> {
+    pub fn render_bundle_encoder_finish<B: GfxBackend>(
+        &self,
+        bundle_encoder: RenderBundleEncoder,
+        desc: &wgt::RenderBundleDescriptor<Label>,
+        id_in: Input<G, id::RenderBundleId>,
+    ) -> id::RenderBundleId {
+        let hub = B::hub(self);
+        let mut token = Token::root();
+        let (device_guard, mut token) = hub.devices.read(&mut token);
+
+        let (data, device_id) = unsafe { bundle_encoder.raw.finish_render() };
+        let device = &device_guard[device_id];
+        let render_bundle = {
+            let (pipeline_layout_guard, mut token) = hub.pipeline_layouts.read(&mut token);
+            let (bind_group_guard, mut token) = hub.bind_groups.read(&mut token);
+            let (pipeline_guard, mut token) = hub.render_pipelines.read(&mut token);
+            let (buffer_guard, _) = hub.buffers.read(&mut token);
+
+            let mut state = State {
+                trackers: TrackerSet::new(device_id.backend()),
+                index: IndexState::new(),
+                vertex: (0..MAX_VERTEX_BUFFERS)
+                    .map(|_| VertexState::new())
+                    .collect(),
+                bind: (0..MAX_BIND_GROUPS).map(|_| BindState::new()).collect(),
+                raw_dynamic_offsets: Vec::new(),
+                flat_dynamic_offsets: Vec::new(),
+                used_bind_groups: 0,
+            };
+            let mut commands = Vec::new();
+
+            let mut peeker = data.as_ptr();
+            let raw_data_end = unsafe { data.as_ptr().offset(data.len() as isize) };
+            let mut command = RenderCommand::Draw {
+                vertex_count: 0,
+                instance_count: 0,
+                first_vertex: 0,
+                first_instance: 0,
+            };
+            loop {
+                assert!(
+                    unsafe { peeker.add(RenderCommand::max_size()) <= raw_data_end },
+                    "RenderCommand (size {}) is too big to fit within raw_data",
+                    RenderCommand::max_size(),
+                );
+                peeker = unsafe { RenderCommand::peek_from(peeker, &mut command) };
+
+                match command {
+                    RenderCommand::SetBindGroup {
+                        index,
+                        num_dynamic_offsets,
+                        bind_group_id,
+                        phantom_offsets,
+                    } => {
+                        let (new_peeker, offsets) = unsafe {
+                            phantom_offsets.decode_unaligned(
+                                peeker,
+                                num_dynamic_offsets as usize,
+                                raw_data_end,
+                            )
+                        };
+                        peeker = new_peeker;
+
+                        if cfg!(debug_assertions) {
```

Comment:
Are we sure this can be off in release? Seems pretty UB-ish to use a misaligned offset. Cost of running it shouldn't be too high, I expect offset count to be low and you're mod-ing by a constant, so the division will be completely removed.

Reply:
agreed!